### PR TITLE
Fixes /pathmap for the unix root ("/") directory.  Adds diagnostics for pathmap failed parsing.

### PR DESCRIPTION
**Pull request by TyOverby**
This PR does three things, all of which are related to the `/pathmap` feature.

1. Adds a test for pathmap by using the CallerFilePath attribute.
2. Fixes the pathmap parser which used to parse `"/"` as `""` which it dropped on the floor without reporting any diagnostics.
3. Adds diagnostics to the pathmap parser.  Anything that it can't parse will make the build fail.

### CODE REVIEW

**Comment by dpoeschl** [comment]
@TyOverby timeouts are probably due to your debugger launches. I'll prefix your PR title with [Testing] so we know to ignore Jenkins failures.

**Comment by TyOverby** [comment]
@dpoeschl: Yep, thanks!  I actually just pushed this up to show @gafter.

**Comment by TyOverby** [comment]
@dpoeschl: Changed the name back because it should work now.

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs [hunk 1]
```diff
@@ -185,7 +185,6 @@ public new CSharpCommandLineArguments Parse(IEnumerable<string> args, string bas
 
 #if DEBUG
                     case "attachdebugger":
-                        Debugger.Launch();
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: No, don't delete this!

**Comment by gafter** [comment]
So was the problem a test bug coupled with insufficient diagnostics for a bad command-line?

File: src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs (hunk 1 above)
Comment (TyOverby) [inline src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs]: Oh Oops!  My "replace all debugger.launch" script went too far!

**Comment by TyOverby** [comment]
The problem was that the path `/` was trimmed out [here](https://github.com/dotnet/roslyn/pull/6330/files#diff-97a7a5298fc5036af649f59ff021c791R202) which led to the PathMap being completely empty at runtime.  

This in combination with insufficient diagnostics led to me thinking that the problem was deeper in the pathmap support than it was, so I fixed up the diagnostics as well.  I figure that if anyone writes a `/pathmap:` option, they should be notified if it fails to parse.

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 2]
```diff
@@ -4657,4 +4657,7 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="ERR_SourceFileReferencesNotSupported" xml:space="preserve">
     <value>Source file references are not supported.</value>
   </data>
+  <data name="ERR_InvalidPathMap" xml:space="preserve">
+    <value>The pathmap option was incorrectly formatted</value>
```
Comment (TyOverby) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: Add a period.

File: src/Compilers/Core/Portable/CommandLine/CommonCommandLineParser.cs [hunk 3]
```diff
@@ -196,8 +196,13 @@ internal static string RemoveTrailingSpacesAndDots(string path)
                 var kv = kEqualsV.Split('=');
                 if (kv.Length != 2) continue;
                 var from = PathUtilities.TrimTrailingSeparators(kv[0]);
+
```
Comment (TyOverby) [inline src/Compilers/Core/Portable/CommandLine/CommonCommandLineParser.cs]: Remove this line.

File: src/Compilers/Core/Portable/CommandLine/CommonCommandLineParser.cs [hunk 4]
```diff
@@ -196,8 +196,13 @@ internal static string RemoveTrailingSpacesAndDots(string path)
...
                 var to = PathUtilities.TrimTrailingSeparators(kv[1]);
-                if (from.Length == 0 || to.Length == 0) continue;
+
+                if (from.Length == 0 || (to.Length == 0 && kv[1] != "/"))
+                {
+                    errors.Add(Diagnostic.Create(_messageProvider, _messageProvider.ERR_InvalidPathMap, kEqualsV));
+                };
                 pathMapBuilder.Add(new KeyValuePair<string, string>(from, to));
```
Comment (TyOverby) [inline src/Compilers/Core/Portable/CommandLine/CommonCommandLineParser.cs]: This should probably be in `else`.

File: src/Compilers/VisualBasic/Portable/VBResources.resx [hunk 5]
```diff
@@ -5332,4 +5332,7 @@
   <data name="ERR_DebugEntryPointNotSourceMethodDefinition" xml:space="preserve">
     <value>Debug entry point must be a definition of a method declared in the current compilation.</value>
   </data>
+  <data name="ERR_InvalidPathMap" xml:space="preserve">
+    <value>The pathmap option was incorrectly formatted</value>
```
Comment (TyOverby) [inline src/Compilers/VisualBasic/Portable/VBResources.resx]: Insert a period.

File: src/Compilers/CSharp/Portable/CSharpResources.resx (hunk 2 above)
Comment (TyOverby) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: Done.

File: src/Compilers/Core/Portable/CommandLine/CommonCommandLineParser.cs (hunk 3 above)
Comment (TyOverby) [inline src/Compilers/Core/Portable/CommandLine/CommonCommandLineParser.cs]: Done.

File: src/Compilers/Core/Portable/CommandLine/CommonCommandLineParser.cs (hunk 4 above)
Comment (TyOverby) [inline src/Compilers/Core/Portable/CommandLine/CommonCommandLineParser.cs]: Done.

File: src/Compilers/VisualBasic/Portable/VBResources.resx (hunk 5 above)
Comment (TyOverby) [inline src/Compilers/VisualBasic/Portable/VBResources.resx]: Done.

**Comment by TyOverby** [comment]
@gafter, @dotnet/roslyn-compiler: Please review.

**Comment by TyOverby** [comment]
In particular, I'm not sure I hit all the spots in the code that I need to for diagnostics.

**Comment by TyOverby** [comment]
@MattGertz I think I need your sign off for this one?

File: src/Compilers/Core/Portable/CommandLine/CommonCommandLineParser.cs [hunk 6]
```diff
@@ -197,8 +197,14 @@ internal static string RemoveTrailingSpacesAndDots(string path)
                 if (kv.Length != 2) continue;
                 var from = PathUtilities.TrimTrailingSeparators(kv[0]);
                 var to = PathUtilities.TrimTrailingSeparators(kv[1]);
-                if (from.Length == 0 || to.Length == 0) continue;
-                pathMapBuilder.Add(new KeyValuePair<string, string>(from, to));
+
+                if (from.Length == 0 || (to.Length == 0 && kv[1] != "/"))
+                {
+                    errors.Add(Diagnostic.Create(_messageProvider, _messageProvider.ERR_InvalidPathMap, kEqualsV));
+                } else
```
Comment (gafter) [inline src/Compilers/Core/Portable/CommandLine/CommonCommandLineParser.cs]: The convention is to place `}` and `else` on separate lines.

**Comment by gafter** [comment]
LGTM

File: src/Compilers/Core/Portable/CommandLine/CommonCommandLineParser.cs (hunk 6 above)
Comment (TyOverby) [inline src/Compilers/Core/Portable/CommandLine/CommonCommandLineParser.cs]: Done.

**Comment by TyOverby** [comment]
Ping @dotnet/roslyn-compiler

**Comment by VSadov** [comment]
LGTM

**Comment by TyOverby** [comment]
@Pilchie or @ManishJayaswal, do I need ask-mode sign-off for this?

File: src/Compilers/Core/VBCSCompilerTests/EndToEndDeterminismTest.cs [hunk 7]
```diff
@@ -46,6 +48,9 @@ private byte[] CompileAndGetBytes(string source, string additionalFlags)
             finally
             {
                 File.Delete(srcFile);
+
+                // TODO(tyoverby) Remove this before merging
+                File.Copy(outFile, tempDir.Path + $"/../test.dll-{ new System.Random().Next() }.dll");
```
Comment (Pilchie) [inline src/Compilers/Core/VBCSCompilerTests/EndToEndDeterminismTest.cs]: Does this still need to be removed?

**Comment by Pilchie** [comment]
In general, for stabilization fixes it would be good to avoid the gratuitous whitespace changes.

**Comment by Pilchie** [comment]
Can you describe in more detail what customer scenario is broken here?

**Comment by VSadov** [comment]
Agree with @Pilchie on whitespace. It is easy to miss little things when there are so many changes.

**Comment by TyOverby** [comment]
Any user that wants to map a path to the unix root directory "/" will have their build be non-deterministic and they won't be warned.

**Comment by TyOverby** [comment]
@Pilchie @VSadov Yeah, I'll go back and fix that up.  My vim configuration automatically stripped the ending whitespace, but I just reconfigured it to not do that anymore.

File: src/Compilers/Core/VBCSCompilerTests/EndToEndDeterminismTest.cs (hunk 7 above)
Comment (TyOverby) [inline src/Compilers/Core/VBCSCompilerTests/EndToEndDeterminismTest.cs]: It did;  My mistake for not catching that during my pass through.

**Comment by TyOverby** [comment]
@Pilchie @VSadov 

Whitespace changes are gone.

**Comment by Pilchie** [comment]
Approved, pending build completion, though the scenario in question was close to the bar for me.

**Comment by TyOverby** [comment]
@gafter I added another failure for pathmaps that have the wrong number of equals signs. `/pathmap:foo,bar,baz` now is an error, as is `/pathmap:foo=bar=baz`.

If these changes look good to you, I'll merge this.

**Comment by gafter** [comment]
@TyOverby Looks good.

In the future can you please add functionality in a separate PR so I don't have to review the whole thing from scratch again?

**Comment by TyOverby** [comment]
@gafter: up at the top where it says [commits](https://github.com/dotnet/roslyn/pull/6330/commits) you can click on the [most recent commit](https://github.com/TyOverby/roslyn/commit/d6e749a8fc4fbfca374e3e7227f89a620787cdc3) which is considerably smaller.

**Pull request merged**